### PR TITLE
Machine learning gem

### PR DESCRIPTION
**Pull request by kberg0**
Adds a new MachineLearning GEM to o3de-extras.
This is a very rudimentary ML package, which can train basic classifier networks using stochastic gradient descent and backward propagation. It's currently using ImGui rather than a Qt panel for training models, and not using the full O3DE asset system instead opting for a simple and fast binary save/load system that leverages AzNetworking binary serialization. Quite a bit of additional debugging and hardening and tooling will be necessary, however the system is now functional.

image

image

image

I have included a very basic pre-trained model trained on the mnist data set. This model achieves a 96% accuracy on the training data and a 94% accuracy on the test data.

`[==========] Running 9 tests from 4 test cases.`
`[----------] Global test environment set-up.`
`[----------] 5 tests from MachineLearning_Activations`
`[ RUN      ] MachineLearning_Activations.OneHotArgMax`
`[       OK ] MachineLearning_Activations.OneHotArgMax (0 ms)`
`[ RUN      ] MachineLearning_Activations.TestRelu`
`[       OK ] MachineLearning_Activations.TestRelu (0 ms)`
`[ RUN      ] MachineLearning_Activations.TestSigmoid`
`[       OK ] MachineLearning_Activations.TestSigmoid (0 ms)`
`[ RUN      ] MachineLearning_Activations.TestSoftmax`
`[       OK ] MachineLearning_Activations.TestSoftmax (0 ms)`
`[ RUN      ] MachineLearning_Activations.TestLinear`
`[       OK ] MachineLearning_Activations.TestLinear (0 ms)`
`[----------] 5 tests from MachineLearning_Activations (2 ms total)`
`[----------] 1 test from MachineLearning_LossFunctions`
`[ RUN      ] MachineLearning_LossFunctions.TestMeanSquaredError`
`[       OK ] MachineLearning_LossFunctions.TestMeanSquaredError (0 ms)`
`[----------] 1 test from MachineLearning_LossFunctions (0 ms total)`
`[----------] 2 tests from MachineLearning_Layers`
`[ RUN      ] MachineLearning_Layers.TestConstructor`
`[       OK ] MachineLearning_Layers.TestConstructor (15 ms)`
`[ RUN      ] MachineLearning_Layers.TestForward`
`[       OK ] MachineLearning_Layers.TestForward (0 ms)`
`[----------] 2 tests from MachineLearning_Layers (16 ms total)`
`[----------] 1 test from MachineLearning_MLP`
`[ RUN      ] MachineLearning_MLP.TestGradientCalculations`
`[       OK ] MachineLearning_MLP.TestGradientCalculations (59 ms)`
`[----------] 1 test from MachineLearning_MLP (60 ms total)`
`[----------] Global test environment tear-down`
`[==========] 9 tests from 4 test cases ran. (82 ms total)`
`[  PASSED  ] 9 tests.`
`OKAY AzRunUnitTests() returned 0`

### CODE REVIEW

File: Gems/MachineLearning/.gitignore
Comment (mbalfour-amzn) [inline Gems/MachineLearning/.gitignore:1]: Shouldn't be included in the PR?

File: Gems/MachineLearning/Code/Platform/Android/PAL_android.cmake [hunk 1]
```diff
@@ -0,0 +1,4 @@
+
+set(PAL_TRAIT_MACHINELEARNING_SUPPORTED TRUE)
+set(PAL_TRAIT_MACHINELEARNING_TEST_SUPPORTED TRUE)
+set(PAL_TRAIT_MACHINELEARNING_EDITOR_TEST_SUPPORTED FALSE)
```
Comment (mbalfour-amzn) [inline Gems/MachineLearning/Code/Platform/Android/PAL_android.cmake:4]: Maybe I'm missing something, but it looks like the traits are set the same on all platforms? If so, why have platform-specific build files at all?

File: Gems/MachineLearning/Code/Source/Algorithms/Training.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+                // We reshuffle the training data indices each epoch to avoid patterns in the training data
+                if (m_shuffleTrainingData)
+                {
+                    AZStd::lock_guard lock(m_mutex);
+                    m_trainData.ShuffleSamples();
+                }
+                ++m_currentEpoch;
+
+                // Generally we want to keep monitoring the models performence on both test and training data
```
Comment (mbalfour-amzn) [inline Gems/MachineLearning/Code/Source/Algorithms/Training.cpp:109]:
```suggestion
                // Generally we want to keep monitoring the model's performance on both test and training data
```

File: Gems/MachineLearning/Code/Source/Algorithms/Training.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        void StartTraining();
+        void StopTraining();
+
+        AZStd::atomic<AZStd::size_t> m_currentEpoch = 0;
+        std::atomic<bool> m_trainingComplete = true;
+
+        AZ::ThreadSafeDeque<float> m_testCosts;
+        AZ::ThreadSafeDeque<float> m_trainCosts;
+
+    //private:
```
Comment (mbalfour-amzn) [inline Gems/MachineLearning/Code/Source/Algorithms/Training.h:55]: Either it's private or it's not, but a commented-out private: shouldn't be here. :P

File: Gems/MachineLearning/Code/Source/Debug/MachineLearningDebugTrainingWindow.cpp [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+        if (trainingInstance->m_layerBiases.size() < layerIndex)
+        {
+            trainingInstance->m_layerBiases.resize(layerIndex + 1);
+            trainingInstance->m_layerBiases[layerIndex].Init("Biases", 250, ImGui::LYImGuiUtils::HistogramContainer::ViewType::Histogram, true, 0.0f, 1.0f, ImGui::LYImGuiUtils::HistogramContainer::AutoExpand);
+        }
+
+        //m_selectedModel->GetLayerWeights(layerIter), m_selectedModel->GetLayerBiases(layerIter)
+
+        //trainingInstance->m_layerWeights[layerIndex].Draw(ImGui::GetColumnWidth(), 200.0f);
+        //trainingInstance->m_layerBiases[layerIndex].Draw(ImGui::GetColumnWidth(), 200.0f);
```
Comment (mbalfour-amzn) [inline Gems/MachineLearning/Code/Source/Debug/MachineLearningDebugTrainingWindow.cpp:146]: Commented-out code should get removed

File: Gems/MachineLearning/Code/Source/Debug/MachineLearningDebugTrainingWindow.cpp [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+            //for (AZStd::size_t layerIter = 0; layerIter < m_selectedModel->GetLayerCount(); ++layerIter)
+            //{
+            //    AZStd::fixed_string<64> name;
+            //    name = AZStd::string::format("Layer %u Parameters", static_cast<uint32_t>(layerIter));
+            //    if (ImGui::CollapsingHeader(name.c_str()))
+            //    {
+            //        DrawLayerParameters(trainingInstance, layerIter);
+            //    }
+            //}
+
```
Comment (mbalfour-amzn) [inline Gems/MachineLearning/Code/Source/Debug/MachineLearningDebugTrainingWindow.cpp:369]: Remove commented-out code

File: Gems/MachineLearning/Code/Source/MachineLearningSystemComponent.h [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+#include <AzCore/Component/Component.h>
+#include <MachineLearning/IMachineLearning.h>
+
+namespace MachineLearning
+{
+    class MachineLearningSystemComponent
+        : public AZ::Component
+        , protected MachineLearningRequestBus::Handler
+//        , public AZ::Interface<IMachineLearning>::Registrar
```
Comment (mbalfour-amzn) [inline Gems/MachineLearning/Code/Source/MachineLearningSystemComponent.h:19]: Remove commented-out code

File: Gems/MachineLearning/Code/Source/Models/Layer.cpp [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+            //{
+            //    for (AZStd::size_t j = 0; j < trainingData.m_weightGradients.GetColumnCount(); ++j)
+            //    {
+            //        AZLOG_INFO("Weight %ux%u : %f", i, j, trainingData.m_weightGradients.GetElement(i, j));
+            //    }
+            //}
+            //for (AZStd::size_t i = 0; i < trainingData.m_biasGradients.GetDimensionality(); ++i)
+            //{
+            //    AZLOG_INFO("Bias %u : %f", i, trainingData.m_biasGradients.GetElement(i));
+            //}
```
Comment (mbalfour-amzn) [inline Gems/MachineLearning/Code/Source/Models/Layer.cpp:217]: Remove unused code

File: Gems/MachineLearning/Code/Source/Models/Layer.h [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        //! @return boolean true for success, false for serialization failure
+        bool Serialize(AzNetworking::ISerializer& serializer);
+
+        //! Returns the estimated size required to serialize this layer.
+        AZStd::size_t EstimateSerializeSize() const;
+
+        //! Updates layer internals for it's requested dimensionalities.
+        void OnSizesChanged();
+
+        // These are intentionally left public so that unit testing can exhaustively examine all layer state
```
Comment (mbalfour-amzn) [inline Gems/MachineLearning/Code/Source/Models/Layer.h:62]: FWIW, you could also make them protected and either have a unit testing class that derives from it to allow state introspection or have a unit testing class that's a friend class.

File: Gems/MachineLearning/Code/Source/Models/MultilayerPerceptron.cpp [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+    }
+
+    AZStd::size_t MultilayerPerceptron::GetInputDimensionality() const
+    {
+        return m_activationCount;
+    }
+
+    AZStd::size_t MultilayerPerceptron::GetOutputDimensionality() const
+    {
+        //AZStd::lock_guard lock(m_mutex);
```
Comment (mbalfour-amzn) [inline Gems/MachineLearning/Code/Source/Models/MultilayerPerceptron.cpp:143]: Remove commented code (repeated throughout this file)

File: Gems/MachineLearning/Code/Source/Models/MultilayerPerceptron.h [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        AZStd::string m_name;
+
+        //! The model asset file.
+        AZStd::string m_modelFile;
+
+        //! Optional test and train asset data files.
+        AZStd::string m_testDataFile;
+        AZStd::string m_testLabelFile;
+        AZStd::string m_trainDataFile;
+        AZStd::string m_trainLabelFile;
```
Comment (mbalfour-amzn) [inline Gems/MachineLearning/Code/Source/Models/MultilayerPerceptron.h:84]: I'm really not a fan of the way this circumvents the asset system. It prevents asynchronous loads, hot-reloading, auto-detecting of the asset dependencies for inclusion in seed lists, etc. What's the reasoning behind not making a simple asset handler around the LoadModel code?

Also, because these paths are strings, am I correct in thinking they don't show file picker dialogs via the Edit Context?

File: Gems/MachineLearning/Code/Source/Nodes/SupervisedLearning.cpp [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+#include <AzCore/std/chrono/chrono.h>
+
+namespace MachineLearning
+{
+    INeuralNetworkPtr SupervisedLearning::In
+    (
+        INeuralNetworkPtr Model,
+        ILabeledTrainingDataPtr TrainingData, 
+        ILabeledTrainingDataPtr TestData,
+        //LossFunctions CostFunction,
```
Comment (mbalfour-amzn) [inline Gems/MachineLearning/Code/Source/Nodes/SupervisedLearning.cpp:23]: Remove commented-out code

File: Gems/MachineLearning/Code/Tests/Algorithms/ActivationTests.cpp [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        for (AZStd::size_t iter = 0; iter < output.GetDimensionality(); ++iter)
+        {
+            ASSERT_GE(output.GetElement(iter), 0.0f);
+            ASSERT_LE(output.GetElement(iter), 1.0f);
+        }
+
+        // Additionally, the sum of all the elements should be <= 1, as softmax returns a probability distribution
+        const float totalSum = output.L1Norm();
+        // Between floating point precision and the estimates we use for exp(x), the total sum probability can be slightly greater than one
+        // We add a small epsilon to account for this error
```
Comment (mbalfour-amzn) [inline Gems/MachineLearning/Code/Tests/Algorithms/ActivationTests.cpp:91]: Should L1Norm include a clamp() to prevent the result from going over one?

File: Gems/MachineLearning/Code/Tests/Models/MultilayerPerceptronTests.cpp [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    class MachineLearning_MLP
+        : public UnitTest::LeakDetectionFixture
+    {
+    };
+
+    TEST_F(MachineLearning_MLP, TestGradientCalculations)
+    {
+        // As the computations performed during gradient descent are non-trivial, this unit test carefully replicates the backward propagation example as laid out in this article
+        // https://mattmazur.com/2015/03/17/a-step-by-step-backpropagation-example/
+        // The example is extremely simple, so in the case of a unit test failure, this allows the maintainer to carefully trace unit test execution and compare its output with the article
```
Comment (mbalfour-amzn) [inline Gems/MachineLearning/Code/Tests/Models/MultilayerPerceptronTests.cpp:25]: Nice! :)

File: Gems/MachineLearning/Code/Platform/Android/PAL_android.cmake (hunk 1 above)
Comment (kberg-amzn) [reply Gems/MachineLearning/Code/Platform/Android/PAL_android.cmake:4]: Comes from the GEM template.. it's all created by default for some reason.

File: Gems/MachineLearning/.gitignore
Comment (kberg-amzn) [reply Gems/MachineLearning/.gitignore:1]: good catch, will remove.

File: Gems/MachineLearning/Code/Source/Algorithms/Training.h (hunk 3 above)
Comment (kberg-amzn) [reply Gems/MachineLearning/Code/Source/Algorithms/Training.h:55]: Will be private in the future 😆 I can remove for now..

File: Gems/MachineLearning/Code/Source/Debug/MachineLearningDebugTrainingWindow.cpp (hunk 4 above)
Comment (kberg-amzn) [reply Gems/MachineLearning/Code/Source/Debug/MachineLearningDebugTrainingWindow.cpp:146]: Yeah, to-be-implemented. Monitoring parameters and gradients is super essential to being able to debug and train a model successfully since there are so many parameters that need to be tuned to get proper convergence.

File: Gems/MachineLearning/Code/Source/Models/MultilayerPerceptron.h (hunk 10 above)
Comment (kberg-amzn) [reply Gems/MachineLearning/Code/Source/Models/MultilayerPerceptron.h:84]: Yup, they're just raw strings.. One thing this allowed that turned out to be really useful was I could save a model, try a refinement training session and just observe the models performance on test and train data, and if I didn't get the result expected I could instantly restore and try a different training approach. I'd like to keep that flexibility as it was *super* useful.

File: Gems/MachineLearning/Code/Source/Nodes/SupervisedLearning.cpp (hunk 11 above)
Comment (kberg-amzn) [reply Gems/MachineLearning/Code/Source/Nodes/SupervisedLearning.cpp:23]: This one needs incremental feature improvements from script canvas, once it supports enum inputs I can drop the raw integral pin and make it something actually meaningful to people. Right now they have to double check code and count which integer corresponds to the enum value they want, and then hand enter that integer into the pins value.

File: Gems/MachineLearning/Code/Tests/Algorithms/ActivationTests.cpp (hunk 12 above)
Comment (kberg-amzn) [reply Gems/MachineLearning/Code/Tests/Algorithms/ActivationTests.cpp:91]: That's a good call, way better solution. I'll add that.

Comment (kberg-amzn) [reply Gems/MachineLearning/Code/Tests/Algorithms/ActivationTests.cpp:91]: https://github.com/o3de/o3de/pull/16643

Comment (kberg-amzn) [reply Gems/MachineLearning/Code/Tests/Algorithms/ActivationTests.cpp:91]: I don't know what I was thinking 🤣 It's absolutely wrong to clamp the output of L1Norm. We *actually* want to clamp the result of softmax. I'm going to abandon that L1 norm change as it breaks that function terribly.

Comment (kberg-amzn) [reply Gems/MachineLearning/Code/Tests/Algorithms/ActivationTests.cpp:91]: Also, if I somehow clamped SoftMax (by normalizing it again), that would honestly kind of break this particular test. The SoftMax should generate a probability distribution such that the sum of all probabilities equals to 1.

https://en.wikipedia.org/wiki/Softmax_function

We just want to make sure that the SoftMax math is actually correct by using this test.

File: Gems/MachineLearning/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (lumberyard-employee-dm) [inline Gems/MachineLearning/CMakeLists.txt:1]: This file is missing the O3DE Copyright from it.

File: Gems/MachineLearning/Code/Platform/Android/PAL_android.cmake (hunk 1 above)
Comment (lumberyard-employee-dm) [reply Gems/MachineLearning/Code/Platform/Android/PAL_android.cmake:4]: They are ament to changed by the Gem author to allow turning on/off GoogleTest support on per platform basis.
For example, to turn on test support for Editor Tests support Windows, then updating the trait in PAL_windows.cmake can be done.

File: Gems/MachineLearning/Code/Include/MachineLearning/ILabeledTrainingData.h [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+namespace MachineLearning
+{
+    struct ILabeledTrainingData
+    {
+        AZ_TYPE_INFO(ILabeledTrainingData, "{50DF457E-3EAC-4114-8444-023E64973AD9}");
+
+        virtual ~ILabeledTrainingData() = default;
+
+        //! Loads the indicated label and data files.
+        virtual bool LoadArchive(const AZStd::string& imageFilename, const AZStd::string& labelFilename) = 0;
```
Comment (lumberyard-employee-dm) [inline Gems/MachineLearning/Code/Include/MachineLearning/ILabeledTrainingData.h:23]: This parameters can be AZ::IO::Path to better document them and to get path API functionality
```suggestion
        //! Loads the indicated label and data files.
        virtual bool LoadArchive(const AZ::IO::Path& imageFilename, const AZ::IO::Path& labelFilename) = 0;
```

File: Gems/MachineLearning/Code/Source/Algorithms/Activations.cpp [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+        const AZStd::size_t numElements = vector.GetDimensionality();
+        float maxValue = 0.0f;
+        AZStd::size_t maxIndex = 0;
+        for (AZStd::size_t iter = 0; iter < numElements; ++iter)
+        {
+            if (vector.GetElement(iter) > maxValue)
+            {
+                maxValue = vector.GetElement(iter);
+                maxIndex = iter;
+            }
```
Comment (lumberyard-employee-dm) [inline Gems/MachineLearning/Code/Source/Algorithms/Activations.cpp:44]: I don't know how expensive the call to `GetElement` is, but you can store the value the first time it is called.
```suggestion
            if (float currValue = vector.GetElement(iter); currValue > maxValue)
            {
                maxValue = currValue;
                maxIndex = iter;
            }
```
